# Pharmacogenomics Machine Learning Project
## Notebook 01: Data Exploration

### Purpose
This notebook exlores the ClinPGx datasets that will be used to build a pharmacogenomics machine learning model.

The objects of this notebook are to:
- Understand the structure of each dataset
- Identify potential merge keys
- Examine data quality
- Document missing values
- Identify duplicate records
- Understand relationships between datasets

This notebook is exploratory only.
No data cleaning or modification will be performed here.

### 1. Import Libraries

In [91]:
# Import the pandas library for reading, manipulating, and analyzing tabular data.
import pandas as pd

# Import NumPy for numerical operations.
import numpy as np

# Import Matplotlib for creating graphs and visualizations.
import matplotlib.pyplot as plt

# Import Path from pathlib to build operating system-independent file paths.
from pathlib import Path


### 2. Configure Display Settings

In [92]:
# Display every column in a DataFrame instead of truncating wide tables.
pd.set_option("display.max_columns", None)

# Allow up to 100 rows to be displayed before truncation for more context when examning datasets.
pd.set_option("display.max_rows", 100)

# Set a consistent default figure size for all plots in this notebook.
plt.rcParams["figure.figsize"] = (10, 6)

### 3. Define Data Paths

In [93]:
# Create a Path object that points to the folder containing the raw ClinPGx datasets.
DATA_PATH = Path("../data/raw")

# Create file paths for each ClinPGx dataset.
clinical_file = DATA_PATH / "clinicalVariants.tsv"
drug_file = DATA_PATH / "var_drug_ann.tsv"
phenotype_file = DATA_PATH / "var_pheno_ann.tsv"
functional_file = DATA_PATH / "var_fa_ann.tsv"

### 4. Load Datasets

In [94]:
# Read the Clinical Variants dataset into a pandas DataFrame.
clinical = pd.read_csv(clinical_file, sep="\t")

# Read the Variant Drug Annotation dataset.
drug = pd.read_csv(drug_file, sep="\t")

# Read the Variant Phenotype Annotation dataset.
phenotype = pd.read_csv(phenotype_file, sep="\t")

# Read the Functional Assay Annotation dataset.
functional = pd.read_csv(functional_file, sep="\t")

In [95]:
# Print the dimensions of each dataset: (number of rows, number of columns)
print("Clinical:", clinical.shape)
print("Drug:", drug.shape)
print("Phenotype:", phenotype.shape)
print("Functional:", functional.shape)

Clinical: (5190, 6)
Drug: (12975, 22)
Phenotype: (14490, 25)
Functional: (2153, 23)


### 5. Dataset Summary
&emsp;This information helps determine whether any columns may require cleaning later.\
&emsp;This includes:\
&emsp;&ensp;- number of rows\
&emsp;&ensp;- number of columns\
&emsp;&ensp;- column names\
&emsp;&ensp;- data types\
&emsp;&ensp;- number of non-missing values

In [96]:
# Display summary information about the Clinical Variants dataset.
clinical.info()

<class 'pandas.DataFrame'>
RangeIndex: 5190 entries, 0 to 5189
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   variant            5190 non-null   str  
 1   gene               4936 non-null   str  
 2   type               5190 non-null   str  
 3   level of evidence  5190 non-null   str  
 4   chemicals          5190 non-null   str  
 5   phenotypes         4175 non-null   str  
dtypes: str(6)
memory usage: 243.4 KB


In [97]:
# Display summary information for the Variant Drug Annotation dataset.
drug.info()

<class 'pandas.DataFrame'>
RangeIndex: 12975 entries, 0 to 12974
Data columns (total 22 columns):
 #   Column                                  Non-Null Count  Dtype
---  ------                                  --------------  -----
 0   Variant Annotation ID                   12975 non-null  int64
 1   Variant/Haplotypes                      12975 non-null  str  
 2   Gene                                    12635 non-null  str  
 3   Drug(s)                                 12975 non-null  str  
 4   PMID                                    12975 non-null  int64
 5   Phenotype Category                      12951 non-null  str  
 6   Significance                            12975 non-null  str  
 7   Notes                                   11340 non-null  str  
 8   Sentence                                12975 non-null  str  
 9   Alleles                                 12436 non-null  str  
 10  Specialty Population                    1339 non-null   str  
 11  Metabolizer types         

In [98]:
# Display summary information for the Variant Phenotype Annotation dataset.
phenotype.info()

<class 'pandas.DataFrame'>
RangeIndex: 14490 entries, 0 to 14489
Data columns (total 25 columns):
 #   Column                                          Non-Null Count  Dtype
---  ------                                          --------------  -----
 0   Variant Annotation ID                           14490 non-null  int64
 1   Variant/Haplotypes                              14490 non-null  str  
 2   Gene                                            14074 non-null  str  
 3   Drug(s)                                         13551 non-null  str  
 4   PMID                                            14490 non-null  int64
 5   Phenotype Category                              14358 non-null  str  
 6   Significance                                    14490 non-null  str  
 7   Notes                                           11762 non-null  str  
 8   Sentence                                        14490 non-null  str  
 9   Alleles                                         14005 non-null  str  
 1

In [99]:
# Display summary information for the Functional Assay dataset.
functional.info()

<class 'pandas.DataFrame'>
RangeIndex: 2153 entries, 0 to 2152
Data columns (total 23 columns):
 #   Column                                          Non-Null Count  Dtype
---  ------                                          --------------  -----
 0   Variant Annotation ID                           2153 non-null   int64
 1   Variant/Haplotypes                              2153 non-null   str  
 2   Gene                                            2103 non-null   str  
 3   Drug(s)                                         1488 non-null   str  
 4   PMID                                            2153 non-null   int64
 5   Phenotype Category                              2005 non-null   str  
 6   Significance                                    2153 non-null   str  
 7   Notes                                           1961 non-null   str  
 8   Sentence                                        2153 non-null   str  
 9   Alleles                                         2146 non-null   str  
 10 

#### Observations

&emsp;The Clinical Variants dataset contains 5,190 records and 6 columns, making it the most concise of the four ClinPGx datasets. Rather than storing detailed study information, it summarizes clinicaly important gene-drug relationships and their associated evidence levels. Because this dataset is highly curated, it is likely to serve as the primary source of clinically relevant labels for the machine learning model. However, its limited number of features suggests that it will need to be supplemented with information from the other ClinPGx datasets to create a richer feature set.

&emsp;The Variant Drug Annotation dataset contains substantially more records and many more descriptive columns than the Clinical Variants dataset. The additional information includes publication identifiers, phenotype categories, significance values, metabolizer classifications, and population information. This dataset appears to provide the detailed scientific evidence underlying many of the summarized clinical relationships. These variables may become useful predictive features after careful preprocessing and standardization. Because this dataset contains considerably more metadata, it is expected to require more extensive cleaning than the Clinical Variants dataset.

&emsp; The Variant Phenotype Annotation dataset is the largest of the four datasets. It contains detailed descriptions of observed phenotypic outcomes associated with genetic variants and drug exposure. The larger number of observations suggests that individual gene-drug relationships may appear multiple times across different studies or populations. These repeated observations are expected and should not automatically be interpreted as duplicate errors. Instead, repeated evidence may later be aggregated into summary features such as the number of supporting studies or the consistency of reported outcomes.

&emsp; The Functional Assay dataset contains laboratory-based evidence describing how genetic variants influence biological function. Unlike the phenotype dataset, these records focus on experimental measurements rather than clinical outcomes. Although this dataset is the smallest of the four, it may provide biologically meaningful information that improves model performance by explaining why particular variants influence drug response. Because functional assay ifnormation differs conceptually from clinical observations, careful consideration will be required when incorporating these data into the final feature set.


### 6. Preview Records
This provides a quick visual inspection of the data structure and verifies that the dataset was imported correctly.

In [100]:
# Display the first five rows of the Clinical Variants dataset.
clinical.head()

,variant,gene,type,level of evidence,chemicals,phenotypes
0,"CYP2C9*1, CYP2C9*3, CYP2C9*13",CYP2C9,Metabolism/PK,1A,lornoxicam,NaN
1,rs17376848,DPYD,Toxicity,1A,capecitabine,Neoplasms
2,rs2297595,DPYD,Toxicity,1A,capecitabine,Neoplasms
3,rs1801265,DPYD,Toxicity,1A,capecitabine,Neoplasms
4,"CYP2C19*1, CYP2C19*2, CYP2C19*3",CYP2C19,Metabolism/PK,1A,dexlansoprazole,NaN


In [101]:
# Display the first five rows of the Variant Drug Annotation dataset.
drug.head()

,Variant Annotation ID,Variant/Haplotypes,Gene,Drug(s),PMID,Phenotype Category,Significance,Notes,Sentence,Alleles,Specialty Population,Metabolizer types,isPlural,Is/Is Not associated,Direction of effect,PD/PK terms,Multiple drugs And/or,Population types,Population Phenotypes or diseases,Multiple phenotypes or diseases And/or,Comparison Allele(s) or Genotype(s),Comparison Metabolizer types
0,1451834452,"CYP3A4*1, CYP3A4*17",CYP3A4,nifedipine,15634941,"Other, Metabolism/PK",not stated,in vitro expression of the recombinant CYP3A4*...,CYP3A4 *17 is associated with decreased metabo...,*17,NaN,NaN,Is,Associated with,decreased,metabolism of,NaN,NaN,NaN,NaN,*1,NaN
1,1453073660,rs2909451,DPP4,sitagliptin,39792745,Efficacy,yes,"""Patients with the rs2909451 TT genotype in th...",Genotype TT is associated with decreased respo...,TT,NaN,NaN,Is,Associated with,decreased,response to,NaN,in people with,"Other:Diabetes Mellitus, Type 2",NaN,NaN,NaN
2,1453071582,rs706795,FAIM2,"citalopram, escitalopram, fluoxetine, fluvoxam...",40054571,Efficacy,no,"""We observed nominally significant association...",Allele T is associated with increased response...,T,NaN,NaN,Is,Associated with,increased,response to,or,in people with,Other:Obsessive-Compulsive Disorder,NaN,C,NaN
3,1451138786,rs16918842,OPRK1,heroin,31940240,Dosage,no,No significant difference in allele or genotyp...,Allele T is not associated with dose of heroin...,T,NaN,NaN,Is,Not associated with,NaN,dose of,NaN,in people with,Other:Heroin Dependence,NaN,C,NaN
4,1448257202,"CYP2C9*1, CYP2C9*3",CYP2C9,warfarin,27488176,Dosage,yes,NaN,CYP2C9 *3 is associated with decreased dose of...,*3,NaN,NaN,Is,Associated with,decreased,dose of,NaN,NaN,NaN,NaN,*1/*1,NaN


In [102]:
# Display the first five rows of the Variant Phenotype Annotation dataset.
phenotype.head()

,Variant Annotation ID,Variant/Haplotypes,Gene,Drug(s),PMID,Phenotype Category,Significance,Notes,Sentence,Alleles,Specialty Population,Metabolizer types,isPlural,Is/Is Not associated,Direction of effect,Side effect/efficacy/other,Phenotype,Multiple phenotypes And/or,When treated with/exposed to/when assayed with,Multiple drugs And/or,Population types,Population Phenotypes or diseases,Multiple phenotypes or diseases And/or,Comparison Allele(s) or Genotype(s),Comparison Metabolizer types
0,1449169911,HLA-B*35:08,HLA-B,lamotrigine,29238301,Toxicity,no,The allele was not significant when comparing ...,HLA-B *35:08 is not associated with likelihood...,*35:08,NaN,NaN,Is,Not associated with,NaN,likelihood of,"Side Effect:Maculopapular Exanthema, Side Effe...",or,when treated with,NaN,in people with,Disease:Epilepsy,NaN,NaN,NaN
1,1450813230,rs1801272,CYP2A6,nicotine,22046326,Toxicity,yes,Referred to in the paper as CYP2A6*2.,Genotypes AA + AT are associated with decrease...,AA + AT,NaN,NaN,Are,Associated with,decreased,risk of,Other:Tobacco Use Disorder,NaN,due to,NaN,NaN,NaN,NaN,TT,NaN
2,1453086503,"UGT1A1*1, UGT1A1*28",UGT1A1,sacituzumab govitecan,39528547,Toxicity,yes,"""Genotype-specific analyses revealed higher in...",UGT1A1 *1/*28 + *28/*28 is associated with inc...,*1/*28 + *28/*28,NaN,NaN,Is,Associated with,increased,severity of,Side Effect:Neutropenia,NaN,when treated with,NaN,in people with,Other:Breast Neoplasms,NaN,*1/*1,NaN
3,1451833920,ABCB1 intermediate activity,ABCB1,loperamide,35815036,Toxicity,not stated,A patient with opioid use disorder showed redu...,ABCB1 intermediate activity is associated with...,NaN,NaN,intermediate activity,Is,Associated with,increased,risk of,Side Effect:Opioid-Related Disorders,NaN,when exposed to,NaN,NaN,NaN,NaN,NaN,NaN
4,1453086600,HLA-A*02:01,HLA-A,"ipilimumab, nivolumab, pembrolizumab",40155873,Toxicity,no,"""This study did not find a significant risk as...",HLA-A *02:01 is not associated with increased ...,*02:01,NaN,NaN,Is,Not associated with,increased,likelihood of,Side Effect:adverse events,NaN,when treated with,or,in people with,"Other:Melanoma, Other:Metastatic neoplasm",and,NaN,NaN


In [103]:
# Display the first five rows of the Functional Assay dataset.
functional.head()

,Variant Annotation ID,Variant/Haplotypes,Gene,Drug(s),PMID,Phenotype Category,Significance,Notes,Sentence,Alleles,Specialty Population,Assay type,Metabolizer types,isPlural,Is/Is Not associated,Direction of effect,Functional terms,Gene/gene product,When treated with/exposed to/when assayed with,Multiple drugs And/or,Cell type,Comparison Allele(s) or Genotype(s),Comparison Metabolizer types
0,1451148445,"CYP2C19*1, CYP2C19*17",CYP2C19,normeperidine,30902024,NaN,not stated,"In other in vitro experiments, normeperidine f...",CYP2C19 *17/*17 is associated with increased f...,*17/*17,NaN,in human liver microsomes,NaN,Is,Associated with,increased,formation of,NaN,NaN,NaN,NaN,*1/*1 + *1/*17,NaN
1,1448281185,"CYP2B6*1, CYP2B6*6",CYP2B6,bupropion,27439448,Efficacy,yes,The ratio of hydroxybupropion versus bupropion...,CYP2B6 *1/*1 is associated with increased acti...,*1/*1,NaN,NaN,NaN,Is,Associated with,increased,activity of,CYP2B6,when assayed with,NaN,NaN,*1/*6,NaN
2,1447944475,"CYP2C9*1, CYP2C9*2",CYP2C9,warfarin,25823787,Metabolism/PK,not stated,CYP2C9*1 formed three hydrogen bonds with warf...,CYP2C9 *2 is associated with decreased affinit...,*2,NaN,crystal structure prediction,NaN,Is,Associated with,decreased,affinity to,CYP2C9,of,NaN,NaN,*1,NaN
3,1447944483,"CYP2C9*1, CYP2C9*3",CYP2C9,warfarin,25823787,Metabolism/PK,not stated,CYP2C9*1 formed three hydrogen bonds with warf...,CYP2C9 *3 is associated with decreased affinit...,*3,NaN,crystal structure prediction,NaN,Is,Associated with,decreased,affinity to,CYP2C9,of,NaN,NaN,*1,NaN
4,1451098400,rs72552763,SLC22A1,ranitidine,29236753,Metabolism/PK,yes,"as reported for OCT1*2, with Vmax decreased by...",Allele del is associated with decreased transp...,del,NaN,Cells,NaN,Is,Associated with,decreased,transport of,NaN,NaN,NaN,NaN,GAT,NaN


#### Observations

&emsp;Examining the first few records confirms that all datasets loaded successfully and that the expected columns are present. In all of the datasets, it seems that some variants are labeled based on the haplotype-based star(*) system while others use reference SNP cluster IDs (rIDs) to avoid legacy structural conflicts. This could be an issue if passed directly into the machine learning pipline, so it may be addressed.

### 7. Column Names
Understanding the available columns is essential for designing a future merge strategy.

In [104]:
# Display the column names in the Clinical Variants dataset.
clinical.columns

Index(['variant', 'gene', 'type', 'level of evidence', 'chemicals',
       'phenotypes'],
      dtype='str')

In [105]:
# Display the column names in the Variant Drug Annotation dataset.
drug.columns

Index(['Variant Annotation ID', 'Variant/Haplotypes', 'Gene', 'Drug(s)',
       'PMID', 'Phenotype Category', 'Significance', 'Notes', 'Sentence',
       'Alleles', 'Specialty Population', 'Metabolizer types', 'isPlural',
       'Is/Is Not associated', 'Direction of effect', 'PD/PK terms',
       'Multiple drugs And/or', 'Population types',
       'Population Phenotypes or diseases',
       'Multiple phenotypes or diseases And/or',
       'Comparison Allele(s) or Genotype(s)', 'Comparison Metabolizer types'],
      dtype='str')

In [106]:
# Display the column names in the Variant Phenotype Annotation dataset.
phenotype.columns

Index(['Variant Annotation ID', 'Variant/Haplotypes', 'Gene', 'Drug(s)',
       'PMID', 'Phenotype Category', 'Significance', 'Notes', 'Sentence',
       'Alleles', 'Specialty Population', 'Metabolizer types', 'isPlural',
       'Is/Is Not associated', 'Direction of effect',
       'Side effect/efficacy/other', 'Phenotype', 'Multiple phenotypes And/or',
       'When treated with/exposed to/when assayed with',
       'Multiple drugs And/or', 'Population types',
       'Population Phenotypes or diseases',
       'Multiple phenotypes or diseases And/or',
       'Comparison Allele(s) or Genotype(s)', 'Comparison Metabolizer types'],
      dtype='str')

In [107]:
# Display the column names in the Functional Assay dataset.
functional.columns

Index(['Variant Annotation ID', 'Variant/Haplotypes', 'Gene', 'Drug(s)',
       'PMID', 'Phenotype Category', 'Significance', 'Notes', 'Sentence',
       'Alleles', 'Specialty Population', 'Assay type', 'Metabolizer types',
       'isPlural', 'Is/Is Not associated', 'Direction of effect',
       'Functional terms', 'Gene/gene product',
       'When treated with/exposed to/when assayed with',
       'Multiple drugs And/or', 'Cell type',
       'Comparison Allele(s) or Genotype(s)', 'Comparison Metabolizer types'],
      dtype='str')

#### Observations
&emsp;Several columns appear to represent the same biological concepts across multiple datasets, although their names and formatting may differ. In particular, columns describing genes, variants, drugs, and phenotypes appear repeatedly throughout the ClinPGx resources. These shared concepts will likely become the foundation of the future merge strategy. Additional columns appear to contain metadata describing supporting publications, evidence, populations, metabolizer classifications, and experimental conditions. These variables may become useful features rather than merge keys.

### 8. Missing Values
Missing values may affect future data cleaning, feature engineering, and machine learning.

In [108]:
# Count the number of missing values in the Clinical Variants dataset.
clinical.isnull().sum()

variant                 0
gene                  254
type                    0
level of evidence       0
chemicals               0
phenotypes           1015
dtype: int64

In [109]:
# Count the number of missing values in the Variant Drug Annotation dataset.
drug.isnull().sum()

Variant Annotation ID                         0
Variant/Haplotypes                            0
Gene                                        340
Drug(s)                                       0
PMID                                          0
Phenotype Category                           24
Significance                                  0
Notes                                      1635
Sentence                                      0
Alleles                                     539
Specialty Population                      11636
Metabolizer types                         12170
isPlural                                      0
Is/Is Not associated                          0
Direction of effect                        4673
PD/PK terms                                   3
Multiple drugs And/or                     11188
Population types                           1760
Population Phenotypes or diseases          3274
Multiple phenotypes or diseases And/or    12408
Comparison Allele(s) or Genotype(s)     

In [110]:
# Count the number of missing values in the Variant Phenotype Annotation dataset.
phenotype.isnull().sum()

Variant Annotation ID                                 0
Variant/Haplotypes                                    0
Gene                                                416
Drug(s)                                             939
PMID                                                  0
Phenotype Category                                  132
Significance                                          0
Notes                                              2728
Sentence                                              0
Alleles                                             485
Specialty Population                              12952
Metabolizer types                                 13831
isPlural                                              0
Is/Is Not associated                                  0
Direction of effect                                4268
Side effect/efficacy/other                         4203
Phenotype                                             5
Multiple phenotypes And/or                      

In [111]:
# Count the number of missing values in the Functional Assay dataset.
functional.isnull().sum()

Variant Annotation ID                                0
Variant/Haplotypes                                   0
Gene                                                50
Drug(s)                                            665
PMID                                                 0
Phenotype Category                                 148
Significance                                         0
Notes                                              192
Sentence                                             0
Alleles                                              7
Specialty Population                              2118
Assay type                                        1410
Metabolizer types                                 2140
isPlural                                             0
Is/Is Not associated                                 0
Direction of effect                                331
Functional terms                                     0
Gene/gene product                                  678
When treat

#### Observations

&emsp;Missing values are present in multiple datasets. Many of the missing values are expected. During the data-cleaning phase, each missing field will be evalued individually to determine whether it should be retained, imputed, or excluded from model training.

&emsp;The Clinical Variants dataset has relatively few missing values compared to the other three datasets. Most missing values occur in the phenotypes column. In terms of potential impact, missing phenotype information could reduce the amount of clinically meaningful information avaliable for model training if phenotype becomes a predictive feature. Missing gene names are more concerning because the gene is likely to be part of the primary merge key. Records without a gene identifier may not be mergeable with the other datasets and will require further investigation during data cleaning.

&emsp;In the Variant Drug Annotation dataset, most missing values occur in optional descriptive columns rather than the primary biological identifiers. Many of these fields describe comparison groups, special patient populations, multiple-drug scenarios, and metabolizer classifications. These pieces of information are only relevant for certain pharmacogenomic studies, so their absence is expected in many records. In terms of potential impact, these missing values are unlikely to prevent dataset integration because they are not expected to be merge keys. However, they may limit the usefulness of these variables as predictive features unless appropriate strategies are used during feature engineering.

&emsp;The Variant Phenotype Annotation dataset contains the largest number of missing values overall, but they concentrated in optional annotation fields. This suggests the database was designed to support many different types of studies. Not every study records every possible variable. In terms of potential impact, many of the missing values likely represent fields that were simply not applicable to a particular publication. However, before training the model, it will need to be determined whether these variables should be excluded, treated as "Unknown" categories, or retained only for records where they are informative.

&emsp;Most missing values in the Functional Assay Annotations occur in experimental metadata rather than the core biological information. Fields such as assay type, cell type, and treatment conditions are naturally absent for some functional studies. In terms of potential impact, these variables may still be useful features for certain records, but because they are missing so frequently, they should not become essential inputs for the machine learning model. Instead, they may be incorporated selectively if they improve predictive performance.

&emsp;Overall, across all four ClinPGx datasets, the majority of missing values occur in optional annotation fields rather than the primary biological identifiers. Variables describing metabolizer classifications, specialty populations, comparison groups, and multiple-drug scenarios are frequently incomplete because they are only relevant to specific study designs. The missingness is systematic rather than random. The relatively low number of missing values in key identifiers such as genes suggests that the datasets should still be suitable for integration into a master dataset.

### 9. Duplicate Records
At this stage, duplicates are only identified, not removed.

In [112]:
# Count the number of duplicate rows in the Clinical Variants dataset.
clinical.duplicated().sum()

np.int64(21)

In [113]:
# Count the number of duplicate rows in the Variant Drug Annotation dataset.
drug.duplicated().sum()

np.int64(0)

In [114]:
# Count the number of duplicate rows in the Variant Phenotype Annotation dataset.
phenotype.duplicated().sum()

np.int64(0)

In [115]:
# Count the number of duplicate rows in the Functional Assay dataset.
functional.duplicated().sum()

np.int64(0)

#### Observations
&emsp;The number of duplicate rows provides an initial indication of whether identifical records exist within each dataset. However, duplicate rows in pharmacogenomics databases require careful interpretation. Multiple records may represent repeated observations from independent scientific studies rather than accidental duplication. Consequently, duplicate records will not be removed during exploration. Their biological meaning must first be investigated during the data integration phase.

### 10. Unique Values
This provides an initial understanding of the genetic diversity contained in the data.

In [ ]:
# Count the number of unique genes represented in the Clinical Variants dataset.
clinical["gene"].nunique()

876

In [ ]:
# Count the number of unique drugs represented in the Clinical Variants dataset.
clinical["chemicals"].nunique()

In [ ]:
# Count the number of unique phenotype descriptions in the Clinical Variants dataset.
clinical["phenotypes"].nunique()

In [ ]:
# Count the number of unique genes represented in the Variant Drug Annotation dataset.
drug["Gene"].nunique()

In [ ]:
# Count the number of unique drugs represented in the Variant Drug Annotation dataset.
drug["Drug(s)"].nunique()

In [ ]:
# Count the number of unique phenotype descriptions in the Variant Drug Annotation datset.
drug["Population Phenotypes of diseases"].nunique()

In [ ]:
# Count the number of unique genes represented in the Variant Phenotype Annotation dataset.
phenotype["Gene"].nunique()

In [ ]:
# Count the number of unique drugs represented in the Variant Phenotype Annotation dataset.
phenotype["Drug(s)"].nunique()

In [ ]:
# Count the number of unique phenotype descriptions in the Variant Phenotype Annotation dataset.
phenotype["Phenotype"].nunique()

In [ ]:
# Count the number of unique genes represented in the Functional Assay dataset.
functional["Gene"].nunique()

In [118]:
# Count the number of unique drugs represented in the Functional Assay dataset.
functional["Drug(s)"].nunique()

295

In [ ]:
# Count the number of unique phenotype descriptions in the Functional Assay dataset.
functional["Phenotype Category"].nunique()

#### Obervations
&emsp;Counting the number of unique genes, drugs, and phenotypes provides an overview of the diversity represented within the ClinPGx datasets. Higher numbers of unique values generally indicate broader biological coverage, while lower numbers may indicate that certain genes or medications are studied extensively. Understanding these distributions will help guide later feature engineering and model evaluation.